### PR TITLE
P1467R9 Extended floating-point types and standard names

### LINKED ISSUE #5579: [2022-07 CWG Motion 14] P1467R9 Extended floating-point types and standard names
[P1467R9](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1467r9.html)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5579
Fixes cplusplus/papers#227

### CODE REVIEW

File: source/numerics.tex
Comment (JohelEGP) [inline source/numerics.tex:9341]:
```suggestion
  // \ref{sf.cmath.cyl.neumann}, cylindrical Neumann functions
```

File: source/numerics.tex
Comment (JohelEGP) [inline source/numerics.tex:9481]:
```suggestion
constexpr @\placeholder{floating-point-type}@ abs(@\placeholder{floating-point-type}@ x);
```

File: source/numerics.tex
Comment (JohelEGP) [inline source/numerics.tex:464]:
```suggestion
the imaginary part with \tcode{other.imag()}.
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/basic.tex
Comment (JohelEGP) [reply source/basic.tex:5530]: After looking further into the index, that is not necessarily the case. Feel free to ignore the comment above.

File: source/preprocessor.tex
Comment (jensmaurer) [reply source/preprocessor.tex:1764]: I like the suggestion, but that seems to be a deviation from the paper. Leaving for @tkoeppe.

File: source/support.tex
Comment (Mick235711) [inline source/support.tex:30]: I don't think there is a pre-existing example of a "Header <xxx> synopsis" being at top-level just under a main library clause (and thus appear in this table), though that's what the paper asked... (which also makes this table looks particularly weird)
Maybe we should still do a [stdfloat] "Floating-point types" with [stdfloat.syn] as subclause (and maybe a one-line general like [cstdint])? Or should we open a new convention here so that in future synopsis can directly appear under a clause?

Comment (JohelEGP) [reply source/support.tex:30]: > I don't think there is a pre-existing example of a "Header <xxx> synopsis" being at second-level just under a main library clause (and thus appear in this table)

You're right.

File: source/preprocessor.tex
Comment (tkoeppe) [reply source/preprocessor.tex:1768]: I'd prefer handling this as a separate issue. Could you open an editorial issue please and we can get feedback from the authors, too?

File: source/support.tex
Comment (tkoeppe) [reply source/support.tex:30]: I agree. We should either start a new subsection "Floating-point types" with additional wording, perhaps relating to C, or otherwise perhaps we can rename "Integer types" to "Numeric types" (or "Arithmetic types") and have both <cstdint> and <stdfloat> in it. I'll open an issue,

File: source/preprocessor.tex
Comment (JohelEGP) [reply source/preprocessor.tex:1768]: Will do.

File: source/iostreams.tex
Comment (tkoeppe) [inline source/iostreams.tex:4745]: Why aren't we using any Elements here? Shouldn't this also be Effects:, and there should be a Returns:?

@dkolsen-pgi?

File: source/numerics.tex
Comment (tkoeppe) [inline source/numerics.tex:1222]: We should probably use `\keyword` for `double` etc., but we can fix that later.

Comment (JohelEGP) [reply source/numerics.tex:1222]: I don't think `\keyword` is used on the library clauses.

File: source/iostreams.tex
Comment (dkolsen-pgi) [reply source/iostreams.tex:4745]: Because none of the other `operator>>` descriptions in [iostream.formatted.arithmetic] use any Elements.  I was just following the existing convention in that section.

**Comment by dkolsen-pgi** [comment]
Section [complex.special] is supposed to be removed in its entirety.  The explicit specializations for `std::complex` are no longer needed with the wording changes in this paper.  There is an easily missed one sentence instruction to remove the section in https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1467r9.html#complex-wording

File: source/numerics.tex
Comment (dkolsen-pgi) [inline source/numerics.tex:9255]: This was the result of a typo when I was doing the mass change to the `<cmath>` synopsis.  The signature of `nexttoward` should be 

`constexpr @\placeholder{floating-point-type}@ nexttoward(@\{placeholder{floating-point-type}@ x, long double y);`

This breaking design change (changing the type of the second parameter of `nexttoward`) was not in the version of the paper that was approved by LEWG.  The change was accidentally introduced without any comment or discussion and without anyone noticing during processing by LWG.

I could fix this with a NB comment.  But I think the mistake is obvious enough that it could be fixed here before the CD is finalized, especially since the author of the mistake swears that it was an honest mistake that no one noticed during the review.

File: source/numerics.tex
Comment (dkolsen-pgi) [inline source/numerics.tex:9287]: The second parameter of `isgreater` should be `@\placeholder{floating-point-type}@ y`, not `double y`.

File: source/support.tex
Comment (dkolsen-pgi) [reply source/support.tex:30]: After shuffling words around in an attempt to make things more clear (with most of the words ending up in [basic.extended.fp]), it ended up with the header `<stdfloat>` being defined entirely by its synopsis.  There were no English words left in the `<stdfloat>` section.  I found this weird, but I didn't know what to do about it.  LWG thought it was okay and didn't ask me to change it.

I have no objection to creating an almost empty section above [stdfloat.syn] so that `<stdfloat>` is organized more like all the other headers.

File: source/support.tex
Comment (dkolsen-pgi) [inline source/support.tex:1445]: That's awkward English.  A phrase with "or" is usually considered to be single, but "the types" is plural.  (That's copied from P1467, so that's my fault for writing that.)

I suggest changing "for the types" to "for any of the types", or changing "or" to "and".  Either of those changes makes the English sound better.  I like the first one ("for any of the types") better.  It should be fine to make this change editorially since it is in a note and doesn't change the meaning of the sentence.

File: source/numerics.tex
Comment (jensmaurer) [reply source/numerics.tex:9255]: The previous comment misses that a paper is no longer "the author's" once is was approved by the various WG21 stages. @tkoeppe , for you to decide.

**Comment by jensmaurer** [comment]
@tkoeppe, for your convenience, I've pushed the changes discussed by @dkolsen-pgi , except for `nexttoward`.

File: source/support.tex
Comment (tkoeppe) [reply source/support.tex:30]: Yes, it's not entirely clear, but we'll take a look after the merging. I'm curious in particular what C is going to do and how this would look after we rebase on C23.

File: source/iostreams.tex
Comment (tkoeppe) [reply source/iostreams.tex:4745]: Interesting, thanks. Let me open an issue for that, but this certainly makes sense as is.

File: source/numerics.tex
Comment (tkoeppe) [reply source/numerics.tex:9165]: Please file an LWG issue for this asap. If we can get LWG sign-off that this was an obvious oversight that can be fixed editorially ("because the intention was clear"), then maybe we can fix this for the CD, otherwise we'll have to fix it through the regular issues process. I'm not sure you even need an NB comment, since we have a standing comment to "fix all issues" anyway, but we should make sure to keep track of this.

File: source/numerics.tex
Comment (tkoeppe) [reply source/numerics.tex:1132]: At least not with any form of consistency. @jensmaurer Do you recall if we have a policy/plan for using `\keyword` in the library?

(No immediate action needed here.)

File: source/numerics.tex
Comment (Dani-Hub) [reply source/numerics.tex:9165]: BTW @tkoeppe : Is it possible to generate a public accessible link to a PDF of the working paper in flight for issues like this one, whose contents cannot reasonably be delta-ed against N4910 and provide this link to me?

File: source/numerics.tex
Comment (tkoeppe) [reply source/numerics.tex]: > Section [complex.special] is supposed to be removed in its entirety. The explicit specializations for `std::complex` are no longer needed with the wording changes in this paper. There is an easily missed one sentence instruction to remove the section in https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1467r9.html#complex-wording

Thank you --  I _really_ wish all wording would be presented in underline/strikethrough and coloured diffs, _all_ of it. For large changes it's OK to just show a short fragment and them say "and 20 more similar cases" or use "[...]" for the middle bits, but having a visual diff marker per edit is immensely valuable.

Thanks also, Jens!

**Comment by tkoeppe** [comment]
> @tkoeppe, for your convenience, I've pushed the changes discussed by @dkolsen-pgi , except for `nexttoward`.

Thank you so much!

File: source/support.tex
Comment (tkoeppe) [reply source/support.tex:1445]: Thanks, good call!

File: source/numerics.tex
Comment (tkoeppe) [reply source/numerics.tex:9165]: @Dani-Hub Hi, thanks! I was just going to contact you and mention the uploaded artefacts we provide for _every_ precommit check. For example, for Motion 13, see https://github.com/cplusplus/draft/actions/runs/2806916819, and download "draft-snapshot". In general, for any commit with a precommit check, click the green tick to see the action run, then go to "Summary", and you should find the PDF there.

Does that help/work for you?

Comment (Dani-Hub) [reply source/numerics.tex:9165]: Yes, thank you very much.

Comment (tkoeppe) [reply source/numerics.tex:9165]: My great pleasure!

File: source/numerics.tex
Comment (jensmaurer) [reply source/numerics.tex:1132]: > Do you recall if we have a policy/plan for using \keyword in the library?

We haven't discussed this in any detail, but if the plan for C++ is approximately that of C in that we want all mentions of keywords colored a little bit (or so), then we should use `\keyword` everywhere in prose text and configure `codeblock` accordingly.

Currently, we really want all keywords in the core section to be annotated, because there will be index entries for them.

File: source/preprocessor.tex
Comment (JohelEGP) [reply source/preprocessor.tex:1768]: The author's suggestion from https://github.com/cplusplus/draft/issues/5715#issuecomment-1207321108.
```suggestion
if and only if the implementation supports an extended floating-point type
with the properties of the \grammarterm{typedef-name} \tcode{std::bfloat16_t} described in \ref{basic.extended.fp}.
```

**Pull request merged**
